activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb

```diff
@@ -560,6 +568,12 @@ def single_value_from_rows(rows)
           row && row.first
         end
 
+        def last_inserted_values(result)
+          return {} if result.rows.empty?
+
+          Hash[result.columns.zip(result.rows.first).map { |key, value| [key, value] }]
```

Comment:
Fwiw this is inefficient from an object allocation pov.  It's probably not great in a hotspot

Reply:
Simply `result.first` will work (from ActiveRecord::Result). It is already optimized as much as possible.